File: Recursion/Queen/Queens.jl
function queens(n)
    solutions = []
    function put!next(state = [])
        next_col = length(state) + 1
        if next_col > n
            solutions = [solutions;[state]]
            return
        end
        function conflict(target_row)
            for (col, row) ∈ enumerate(state)
                if abs(row - target_row) ∈ (0, next_col - col)
                    return true
                end
            end
            return false
        end
        for row ∈ 1:n
            conflict(row) || put!next([state;[row]])
        end
    end
    put!next()
    return solutions
end

for q ∈ queens(4)
    println(q)
end

File: Recursion/Queen/Queens.py
def queens(n):
    def next(state=()):
        next_col = len(state)

        def conflict(target_row):
            for col, row in enumerate(state):
                if abs(row-target_row) in (0, next_col-col):
                    return True
            return False

        for row in range(n):
            if not conflict(row):
                if next_col is n-1:
                    yield state+(row,)
                    return
                for result in next(state+(row,)):
                    yield result

    return next()


for solution in queens(8):
    print(solution)
